akka-projection-core/src/test/scala/akka/projection/TestStatusObserver.scala

```diff
@@ -29,7 +33,9 @@ object TestStatusObserver {
 class TestStatusObserver[Envelope](
     probe: ActorRef[TestStatusObserver.Status],
     lifecycle: Boolean = false,
-    progressProbe: Option[ActorRef[TestStatusObserver.Progress[Envelope]]] = None)
+    offsetProgressProbe: Option[ActorRef[TestStatusObserver.OffsetProgress[Envelope]]] = None,
+    beforeEnvelopeProbe: Option[ActorRef[TestStatusObserver.Before[Envelope]]] = None,
+    afterEnvelopeProbe: Option[ActorRef[TestStatusObserver.After[Envelope]]] = None)
```

Comment:
the probes for `before` and `after` must be separate because the order in which these two actions happen is non-deterministic.

Reply:
Do mean that `before` can be called twice or more before the first element reaches `after`? Is that why you say is non-deterministic? 

Makes sense to have separate probes.

Reply:
What I meant is that we can only guarantee that `before(2)` happens before `before(3)`. But in that case, we can't always guarantee when `after(1)` of `after(2)` will happen.

Also, there are several tests in which the invocations to `before` really become non-deterministic. And also, there is at least one tests in which the invocations to `after` also become non-deterministic due to the race between stream cancelation and inflight elements in the stream.